pkg/kubectl/cmd/util/cmdtools.go

```diff
@@ -0,0 +1,265 @@
+				return fmt.Errorf("unable to launch the %s %q", t.Name, strings.Join(cmdPrefix, " "))
+			}
+		}
+
+		//diff command will return 1 when compare different
+		if t.Name == "diff" {
+			if _, ok := err.(*exec.ExitError); ok {
+				return nil
+			}
+		}
+
+		return fmt.Errorf("unknown error with the %s %q", t.Name, strings.Join(cmdPrefix, " "))
+	}
+	return nil
+}
+
+type TempFile struct {
+	Prefix string
+	Suffix string
+	Buffer io.Reader
+}
+
+// LaunchTempFiles reads the provided stream into multiple temporary file in the given directory
+// and file prefix, and then invokes Launch with the path of that file. It will return
+// the contents of the first file after launch, any errors that occur, and the path of the
+// temporary file so the caller can clean it up as needed.
+func (t CmdTool) LaunchTempFiles(tmpFile ...TempFile) ([]string, error) {
+	var paths []string
+	for _, f := range tmpFile {
+		path, err := prepareTempFile(f.Prefix, f.Suffix, f.Buffer)
+		if err != nil {
+			return nil, err
+		}
+		paths = append(paths, path)
+	}
+	return paths, t.Launch(paths)
+}
+
+// LaunchTempFile reads the provided stream into a temporary file in the given directory
+// and file prefix, and then invokes LaunchTempFiles return the path of that file. It will return
+// the contents of the file after launch, any errors that occur, and the path of the
+// temporary file so the caller can clean it up as needed.
+func (t CmdTool) LaunchTempFile(tmpFile TempFile) ([]byte, string, error) {
+	paths, err := t.LaunchTempFiles(tmpFile)
+	if err != nil {
+		return nil, "", err
+	}
+	bytes, err := ioutil.ReadFile(paths[0])
+	return bytes, paths[0], err
+}
+
+// prepareTempFile create a generated temp file and write stream content into this file
+func prepareTempFile(prefix, suffix string, r io.Reader) (string, error) {
+	f, err := tempFile(prefix, suffix)
+	if err != nil {
+		return "", err
+	}
+	defer f.Close()
+	path := f.Name()
+	if _, err := io.Copy(f, r); err != nil {
+		os.Remove(path)
+		return "", err
+	}
+	// This file descriptor needs to close so the next process (Launch) can claim it.
+	f.Close()
+	return path, nil
+}
+
+// tempFile is a reinvent the wheel of ioutil.TempFile. because ioutil.TempFile does not support extensions for temp files,
+// which means temporary files on platforms with extensions do not perform properly. see https://github.com/openshift/origin/pull/1795
+func tempFile(prefix, suffix string) (f *os.File, err error) {
+	dir := os.TempDir()
+
+	for i := 0; i < 10000; i++ {
+		name := filepath.Join(dir, prefix+randSeq(5)+suffix)
+		f, err = os.OpenFile(name, os.O_RDWR|os.O_CREATE|os.O_EXCL, 0600)
+		if os.IsExist(err) {
+			continue
+		}
+		break
+	}
+	return
+}
+
+var letters = []rune("abcdefghijklmnopqrstuvwxyz0123456789")
+
+func randSeq(n int) string {
+	b := make([]rune, n)
+	for i := range b {
+		b[i] = letters[rand.Intn(len(letters))]
+	}
+	return string(b)
+}
+
+func platformize(linux, windows string) string {
```

Comment:
I think I had a comment about how this should be `unix` rather than `linux`, due to our support of MacOS.